Add face landmark task

vinlemon opened the pull request:
Fixes #7 

This PR adds the Face Landmark task #7 .
The handling of `output_face_blendshapes` and `output_facial_transformation_matrixes` is still missing.

hydai reviewed (changes requested):
As I am not familiar with the face landmark task, I have a few questions, could you please help me to figure out?

hydai commented:
Also, would you like to add a new example under README to show this fantastic feature?

apepkuss commented:
@vinlemon Could you please provide an example in a markdown file, showing how to use the code to finish a face landmark task? The example can help users get familiar with your code and how to use them much more efficient. Thanks a lot!

MediosZ commented on src/tasks/vision/face_landmark/mod.rs:
```diff
@@ -0,0 +1,196 @@
+        );
+
+        let mut tensors_to_landmarks =
+            TensorsToLandmarks::new(468, landmarks_out, landmarks_shape)?;
```
468 seems like the total number of detected facial landmarks

vinlemon replied:
The number 468 corresponds to the standard number of facial landmarks used in MediaPipe's Face Mesh model (kMeshLandmarksNum) as defined [here](https://github.com/google/mediapipe/blob/master/mediapipe/tasks/cc/vision/face_landmarker/tensors_to_face_landmarks_graph.cc). For models including the iris, it increases to 478.

hydai replied:
I see. Could you please add this reference to the comments for it? It would be helpful for those who don't have much background knowledge.

vinlemon replied:
done

MediosZ commented on src/tasks/vision/face_landmark/face_landmark_connections.rs:
```diff
@@ -0,0 +1,2884 @@
+    pub fn get_connections(&self) -> &'static [(usize, usize)] {
+        match self {
+            FaceLandmarkConnections::FacemeshLips => &[
+                (61, 146),
```
Here's a reference:
https://github.com/google-ai-edge/mediapipe/blob/master/mediapipe/python/solutions/face_mesh_connections.py
I guess we may find a paper for this eventually

MediosZ commented on src/tasks/vision/face_landmark/mod.rs:
```diff
@@ -0,0 +1,196 @@
+}
+
+impl FaceLandmarker {
+    const LANDMARKS_NORMALIZE_Z: f32 = 0.4;
```
According to [the proto file](https://github.com/google-ai-edge/mediapipe/blob/c67c7f716455fc91c04ab597d5f90521fde5976c/mediapipe/calculators/tflite/tflite_tensors_to_landmarks_calculator.proto#L58), the default value is 1.0, 
I suggest setting it to 1.0

vinlemon commented on scripts/download-models.sh:
```diff
@@ -39,8 +39,13 @@ image_classification_init() {
+  # for custom model downloaded from kaggle
+  bird_model_name="lite-model_aiy_vision_classifier_birds_V1_3"
+  kaggle_tflite_filename="3.tflite"
+  curl -sL "https://www.kaggle.com/api/v1/models/google/aiy/tfLite/vision-classifier-birds-v1/3/download" -o "${bird_model_name}.tar.gz"
```
The reason for using the Kaggle website instead of the TensorFlow Hub is that TensorFlow Hub has migrated to Kaggle Models, as indicated [here](https://www.kaggle.com/tfhub-dev-faqs). This change caused the previous test to fail, so I updated the source to download the model from Kaggle directly.

vinlemon commented on src/tasks/vision/face_landmark/mod.rs:
```diff
@@ -0,0 +1,196 @@
+                img_h,
+                false,
+            )
+            .transform(img_w, img_h, 1.5, 1.5, 0.0, 0.0, None, false);
```
I took the values of `1.5` from the [MediaPipe repository](https://github.com/google/mediapipe/blob/master/mediapipe/tasks/cc/vision/face_landmarker/face_landmarks_detector_graph.cc) under `ConfigureFaceRectTransformationCalculator`, and they never change in their code.

MediosZ reviewed (approved):
The code LGTM, but I have a comment about the source of the model.
I notice there are two models for face_landmarker, one from [mediapipe-models](https://storage.googleapis.com/mediapipe-models/face_landmarker/face_landmarker/float16/latest/face_landmarker.task) and the other from [mediapipe-tasks](https://storage.googleapis.com/mediapipe-tasks/face_landmarker/face_landmarker.task). They have different number of outputs.
I think we should stick to one of them.
The script downloads the one from mediapipe-tasks but the one from mediapipe-models is more accessible, since it's on the [official website](https://ai.google.dev/edge/mediapipe/solutions/vision/face_landmarker)

vinlemon commented on src/tasks/vision/face_landmark/builder.rs:
```diff
@@ -0,0 +1,107 @@
+        let model_resource = crate::model::parse_model(landmark_file.as_ref())?;
+
+        // check model
+        model_base_check_impl!(model_resource, 1, 2);
```
I used `mediapipe-tasks` because it shares the same download path as the other models, but I agree that switching to `mediapipe-models` is a good idea since it's more up-to-date. This will require a few changes, so I think it makes sense to do it in a new PR, where I'll also add the blendshapes handling.

hydai replied:
I agree. Let's have another PR for this.

vinlemon commented:
@hydai @apepkuss  I've updated the README and added a separate file for the example, which should make the face landmark task clear and easy to follow. Let me know if anything else is needed.

hydai reviewed (approved):
LGTM. @apepkuss PTAL. If everything is good, then let's merge it :-)

hydai commented:
Thanks @vinlemon
